main/src/ca/uwaterloo/flix/language/phase/unification/BddFormula.scala

```diff
@@ -15,10 +15,166 @@
  */
 package ca.uwaterloo.flix.language.phase.unification
 
+import ca.uwaterloo.flix.language.ast.{SourceLocation, Symbol, Type}
+import ca.uwaterloo.flix.util.InternalCompilerException
+import ca.uwaterloo.flix.util.collection.Bimap
+
+import scala.collection.immutable.SortedSet
+import org.sosy_lab.pjbdd.api.{Builders, DD}
+
 object BddFormula {
 
-  class Bdd
+  val creator = Builders.bddBuilder().build()
+
+  class BddFormula(val dd: DD) extends Formula { }
+
+  implicit val AsBoolAlg: BoolAlg[BddFormula] = new BoolAlg[BddFormula] {
+    /**
+      * Returns `true` if `f` represents TRUE.
+      */
+    override def isTrue(f: BddFormula): Boolean = f.dd.isTrue()
+
+    /**
+      * Returns `true` if `f` represents FALSE.
+      */
+    override def isFalse(f: BddFormula): Boolean = f.dd.isFalse()
+
+    /**
+      * Returns `true` if `f` represents a variable.
+      */
+    override def isVar(f: BddFormula): Boolean =
+      f.dd.equalsTo(f.dd.getVariable, creator.makeFalse(), creator.makeTrue())
+
+    /**
+      * Returns a representation of TRUE.
+      */
+    override def mkTrue: BddFormula = {
+      new BddFormula(creator.makeTrue())
+    }
+
+    /**
+      * Returns a representation of FALSE.
+      */
+    override def mkFalse: BddFormula = {
+      new BddFormula(creator.makeFalse())
+    }
+
+    /**
+      * Returns a representation of the variable with the given `id`.
+      */
+    override def mkVar(id: Int): BddFormula = {
+      new BddFormula(creator.makeIthVar(id))
+    }
+
+    /**
+      * Returns a representation of the complement of `f`.
+      */
+    override def mkNot(f: BddFormula): BddFormula = {
+      new BddFormula(creator.makeNot(f.dd))
+    }
+
+    /**
+      * Returns a representation of the disjunction of `f1` and `f2`.
+      */
+    override def mkOr(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeOr(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the conjunction of `f1` and `f2`.
+      */
+    override def mkAnd(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeAnd(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the formula `f1 == f2`.
+      */
+    override def mkXor(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeXor(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns the set of free variables in `f`.
+      */
+    override def freeVars(f: BddFormula): SortedSet[Int] = {
+      freeVarsAux(f.dd)
+    }
+
+    private def freeVarsAux(dd: DD): SortedSet[Int] = {
```

Comment:
Every function should have some documentation. Since we are not (excluding Jaco) experts on BDDs, please add some simple commentary about how the function works.

Reply:
Oh, and any reason this could not be implemented directly in freeVars?

Reply:
No I just thought it was ugly to have to wrap the BDDs in BddFormulas when doing recursive calls